pkg/serviceaccounts/controllers/pull_secrets.go

```diff
@@ -0,0 +1,304 @@
+	secret := obj.(*api.Secret)
+
+	if _, tokenExists := secret.Data[api.ServiceAccountTokenKey]; secret.Type == api.SecretTypeServiceAccountToken && tokenExists {
+		dockercfgSecret, err := e.findDockercfgSecret(secret)
+		if dockercfgSecret == nil || kapierrors.IsNotFound(err) {
+			// we don't have a dockercfgSecret yet, create one
+			dockercfgSecret, err = e.createDockerPullSecret(secret)
+			if err != nil {
+				glog.Error(err)
+				return
+			}
+		}
+
+		for i := 1; i <= NumServiceAccountUpdateRetries; i++ {
+			if err := e.createDockerPullSecretReference(dockercfgSecret); err != nil {
+				if kapierrors.IsConflict(err) && i < NumServiceAccountUpdateRetries {
+					continue
+				}
+
+				glog.Error(err)
+				return
+			}
+
+			// sucess
+			return
+		}
+	}
+}
+
+// secretDeleted reacts to a Secret being deleted by looking to see if it's a dockercfg secret for a service account, in which case it
+// it removes the references from the service account.  Or, if it's a service account token secret, chases the reference to a dockercfg
+// secret, in which case it deletes the dockercfg secret.
+func (e *PullSecretController) secretDeleted(obj interface{}) {
+	secret := obj.(*api.Secret)
+
+	switch {
+	case secret.Type == api.SecretTypeServiceAccountToken:
+		dockercfgSecret, err := e.findDockercfgSecret(secret)
+		if err != nil {
+			glog.Error(err)
+			return
+		}
+		if dockercfgSecret == nil {
+			// no secret to remove
+			return
+		}
+
+		if err := e.client.Secrets(secret.Namespace).Delete(dockercfgSecret.Name); err != nil {
+			glog.Error(err)
+			return
+		}
+
+	case secret.Type == api.SecretTypeDockercfg:
+		for i := 1; i <= NumServiceAccountUpdateRetries; i++ {
+			if _, exists := secret.Annotations[ServiceAccountTokenSecretNameKey]; exists {
+				if _, err := e.removeDockercfgSecretReference(secret); err != nil {
+					if kapierrors.IsConflict(err) && i < NumServiceAccountUpdateRetries {
+						continue
+					}
+
+					glog.Error(err)
+					return
+				}
+
+				// sucess
+				return
+			}
+		}
+	}
+}
+
+// createDockerPullSecret creates a dockercfg secret based on the token secret
+func (e *PullSecretController) createDockerPullSecret(tokenSecret *api.Secret) (*api.Secret, error) {
+	saName, saUID := tokenSecret.Annotations[api.ServiceAccountNameKey], tokenSecret.Annotations[api.ServiceAccountUIDKey]
+
+	// Build the secret
+	dockercfgSecret := &api.Secret{
+		ObjectMeta: api.ObjectMeta{
+			Name:      secret.Strategy.GenerateName(fmt.Sprintf("%s-dockercfg-", saName)),
+			Namespace: tokenSecret.Namespace,
+			Annotations: map[string]string{
+				api.ServiceAccountNameKey:        saName,
+				api.ServiceAccountUIDKey:         saUID,
+				ServiceAccountTokenSecretNameKey: string(tokenSecret.Name),
+			},
+		},
+		Type: api.SecretTypeDockercfg,
+		Data: map[string][]byte{},
+	}
+
+	dockercfg := &credentialprovider.DockerConfig{
+		OpenshiftDockerURL: credentialprovider.DockerConfigEntry{
```

Comment:
@liggitt 

> we need to be able to handle at least the "change" from "it doesn't exist" to "it exists at IP 1.2.3.4", which means we have to rescan all tokens at the point to ensure the registry exists?

You'd really need to list/watch services and react to a change to the docker-registry service, right?